src/ring_theory/integral_domain.lean

```diff
@@ -15,33 +15,60 @@ import algebra.geom_sum
 
 section
 
-open finset polynomial
-
-variables {R : Type*} [integral_domain R] (S : set (units R)) [is_subgroup S] [fintype S]
+open finset polynomial function
+open_locale big_operators nat
 
-lemma card_nth_roots_subgroup_units [decidable_eq R] {n : ℕ} (hn : 0 < n) (a : S) :
-  (univ.filter (λ b : S, b ^ n = a)).card ≤ (nth_roots n ((a : units R) : R)).card :=
-card_le_card_of_inj_on (λ a, ((a : units R) : R))
-  (by simp [mem_nth_roots hn, (units.coe_pow _ _).symm, -units.coe_pow, units.ext_iff.symm, subtype.coe_ext])
-  (by simp [units.ext_iff.symm, subtype.coe_ext.symm])
+variables {R : Type*} {G : Type*} [integral_domain R] [group G] [fintype G]
 
-instance subgroup_units_cyclic : is_cyclic S :=
-by haveI := classical.dec_eq R; exact
-is_cyclic_of_card_pow_eq_one_le
-  (λ n hn, le_trans (card_nth_roots_subgroup_units S hn 1) (card_nth_roots _ _))
+lemma card_nth_roots_subgroup_units (f : G →* R) (hf : injective f) {n : ℕ} (hn : 0 < n) (g₀ : G) :
+  ({g ∈ univ | g ^ n = g₀} : finset G).card ≤ (nth_roots n (f g₀)).card :=
+begin
+  apply card_le_card_of_inj_on f,
+  { intros g hg, rw [sep_def, mem_filter] at hg, rw [mem_nth_roots hn, ← f.map_pow, hg.2] },
+  { intros, apply hf, assumption }
+end
 
+/-- A finite subgroup of the unit group of an integral domain is cyclic. -/
+lemma is_cyclic_of_subgroup_integral_domain (f : G →* R) (hf : injective f) : is_cyclic G :=
+begin
+  haveI := classical.dec_eq G,
+  apply is_cyclic_of_card_pow_eq_one_le,
+  intros n hn,
+  convert (le_trans (card_nth_roots_subgroup_units f hf hn 1) (card_nth_roots n (f 1)))
 end
 
+/-- The unit group of a finite integral domain is cyclic. -/
+instance [fintype R] : is_cyclic (units R) :=
+is_cyclic_of_subgroup_integral_domain
+  { to_fun := coe, map_one' := units.coe_one, map_mul' := units.coe_mul } $ units.ext
```

Comment:
Aha, I think you interpreted what I said as something else good, but I just meant that `units R ->* R` should probably exist as a top-level definition (I think that's what you have inlined here as `{ ... }`), and live wherever the units of a monoid are defined.

Reply:
Aha, I see. What would be a good name? `val_hom`?

Reply:
Done.